Modify rule S2232: reworked rule description for LayC format [SONARJAVA-4490] (#2580)
SonarSource/rspec: rules/S2232/java/rule.adoc

== Why is this an issue?

> The `ResultSet.isLast()` method in package `java.sql` returns `true` if the cursor is on the last row of the `ResultSet`.

Reviewer: ```suggestion
```

Author: Fixed.

There are several reasons to avoid using this method:

1. It is optionally available only for result sets of type `ResultSet.TYPE_FORWARD_ONLY`.
   Database drivers will throw an exception if not supported.
2. The method can be expensive to execute as the database driver may need to fetch ahead one row to determine whether the current row is the last in the result set.
   The documentation of the method explicitly mentions this fact.
3. What "the cursor is on the last row" means for an empty `ResultSet` is unclear.
   Database drivers may return `true` or `false` in this case .

`ResultSet.next()` is a good alternative to `ResultSet.isLast()` as it does not have the mentioned issues.
It is always supported and, as per specification, returns `false` for empty ResultSets.

== How to fix it

Replace uses of `ResultSet.isLast()` with `ResultSet.next()`.

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
ResultSet rs = stmt.executeQuery("SELECT name, address FROM PERSON");

> while (!rs.isLast()) { // Noncompliant

Reviewer: // ..

StringBUidler sb 
```java
while (rs.next() && !rs.isLast()) {
   sb.append(rs.getString("name") + ", ");
}

sb,append(rs.getString("name"))

```

  // process row
}
----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----
ResultSet rs = stmt.executeQuery("SELECT name, address FROM PERSON");
while (rs.next()) { // Compliant
  // process row

> }

Reviewer: ```java
List<String> sb;
while (rs.next()) {
  sb.add(rs.getString("name");
}
sb.stream().collect(Collectors.joining(","));
````

Author: Fixed.

There are several reasons to avoid using this method:

1. It is optionally available only for result sets of type `ResultSet.TYPE_FORWARD_ONLY`.
   Database drivers will throw an exception if not supported.
2. The method can be expensive to execute as the database driver may need to fetch ahead one row to determine whether the current row is the last in the result set.
   The documentation of the method explicitly mentions this fact.
3. What "the cursor is on the last row" means for an empty `ResultSet` is unclear.
   Database drivers may return `true` or `false` in this case .

`ResultSet.next()` is a good alternative to `ResultSet.isLast()` as it does not have the mentioned issues.
It is always supported and, as per specification, returns `false` for empty result sets.

== How to fix it

Refactor your code to use `ResultSet.next()` instead of `ResultSet.isLast()`.
Be cautious of its different semantics and side effects on cursor positioning in the result set.
Verify that your program logic is still valid under these side effects and otherwise adjust it.

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----

> ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
> StringBuilder sb = new StringBuilder();
> results.next();
> while (!results.isLast()) { // Noncompliant
>   sb.append(results.getString("name") + ", ");
>   results.next();
> }
> sb.append(results.getString("name"));

Reviewer: ```suggestion
ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
StringBuilder sb = new StringBuilder();
while (!results.isLast()) { // Noncompliant
  sb.append(results.getString("name") + ", ");
  results.next();
}
sb.append(results.getString("name"));
```

> ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
> StringBuilder sb = new StringBuilder();
> results.next();
> while (!results.isLast()) { // Noncompliant
>   sb.append(results.getString("name") + ", ");
>   results.next();
> }
> sb.append(results.getString("name"));

Reviewer: ```suggestion
ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
StringBuilder sb = new StringBuilder();
while (results.next() && !results.isLast()) { // Noncompliant
  sb.append(results.getString("name") + ", ");
}
String formattedNames = sb.append(results.getString("name"));
```

----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----

> ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
> StringBuilder sb = new StringBuilder();
> if (results.next()) {
>   sb.append(results.getString("name"));
>   while (results.next()) { // Compliant, and program logic refactored
>     sb.append(", " + results.getString("name"));
>   }
> }

Reviewer: ```suggestion
ResultSet results = stmt.executeQuery("SELECT name, address FROM PERSON");
List<String> names = new ArrayList<>();
while (results.next()) { // Compliant, and program logic refactored
  names.add(results.getString("name"));
}
String formattedNames =  names.stream().collect(Collectors.joining(", "));
```
